routemaster/state_machine/tests/test_actions.py

```diff
@@ -62,6 +66,104 @@ def test_actions_do_not_advance_state_on_fail(app_config, create_label, mock_web
     ])
 
 
+def test_actions_retries_use_same_idempotency_token(app_config, create_label, mock_webhook, assert_history):
+    state_machine = app_config.config.state_machines['test_machine']
+
+    expected = {'token': None}
+
+    def persist_token(url, content_type, data, token):
+        expected['token'] = token
+        return WebhookResult.FAIL
+
+    # First get the label into the action state by failing the automatic
+    # progression through the machine.
+    with mock.patch(
+        'routemaster.webhooks.RequestsWebhookRunner.__call__',
+        side_effect=persist_token,
+    ) as webhook:
+        create_label('foo', state_machine.name, {'should_progress': True})
+
+    assert expected['token'] is not None
+
+    webhook.assert_called_once_with(
+        'http://localhost/hook',
+        'application/json',
+        b'{"label": "foo", "metadata": {"should_progress": true}}',
+        expected['token'],
+    )
+
+    with mock_webhook(WebhookResult.FAIL) as webhook:
+        process_retries(
+            app_config,
+            state_machine,
+            state_machine.states[1],
+        )
+
+    webhook.assert_called_once_with(
+        'http://localhost/hook',
+        'application/json',
+        b'{"label": "foo", "metadata": {"should_progress": true}}',
+        expected['token'],
+    )
+
+    with mock_webhook(WebhookResult.SUCCESS) as webhook:
+        process_retries(
+            app_config,
+            state_machine,
+            state_machine.states[1],
+        )
+
+    webhook.assert_called_once_with(
+        'http://localhost/hook',
+        'application/json',
+        b'{"label": "foo", "metadata": {"should_progress": true}}',
+        expected['token'],
+    )
+
+
+def test_different_actions_use_different_idempotency_tokens(app_config, create_label, mock_webhook, assert_history):
+    state_machine = app_config.config.state_machines['test_machine']
+
+    seen_tokens = []
+
+    def persist_token(url, content_type, data, token):
+        seen_tokens.append(token)
+        return WebhookResult.SUCCESS
+
+    # First get the label into the action state by failing the automatic
+    # progression through the machine.
+    with mock.patch(
+        'routemaster.webhooks.RequestsWebhookRunner.__call__',
+        side_effect=persist_token,
+    ) as webhook:
+        create_label('foo', state_machine.name, {'should_progress': True})
+        create_label('bar', state_machine.name, {'should_progress': True})
+        create_label('baz', state_machine.name, {'should_progress': True})
```

Comment:
Seems a shame we have to resort to mocking when the webhook runner is something that is passed in. Could we attach it to the app config?

Reply:
This needs more thought as I've now messed up the API with the webhook configs.

I'm going to:

 - Pull the constructor out of the `RequestsWebhookRunner`, making `__call__` instead take an optional dict of headers.
 - Pull the header selection logic out onto the config model objects in a `headers_for_url` method.
 - Add the webhook runner to the app config class.
 - Add some methods to the test app config to control the behaviour and remove the mocking.

Also I want to extend `create_label` in tests to take an optional state, and create the history, but not actually execute the states for the path to that state. This should simplify the tests considerably.

I'll do all of this in a separate PR.

Reply:
#23